### DISCUSSION TITLE
Custom renderer support?

### DISCUSSION THREAD

**Discussion by Josh015** [bevyengine/bevy#1420] [Q&A] [answered]
Is it possible to build a custom renderer using Bevy's APIs? Specifically, I'd like to make a simple deferred renderer which will require MRT and multiple pass support. Is that even possible currently?

I'm still new to Bevy, so forgive me if I've misunderstood.

**EDIT:** To clarify, I mean while working within `bevy_render` rather than completely replacing it and doing a fully custom rendering API.

**Comment by bjorn3** [comment, marked as answer]
# Can you?

Bevy has a very modular design. Almost all components are regular plugins like you could write yourself. You can choose to omit the `bevy_render` and `bevy_wgpu` (a wgpu based backend for `bevy_render`) plugin and write your own renderer instead if you want. To disable `bevy_render`, I think you can use `app.add_plugins_with(DefaultPlugins, |plugins| plugins.disable::<bevy::render::RenderPlugin>().disable::<bevy::wgpu::WgpuPlugin>())` instead of `app.add_plugins(DefaultPlugins)`. You should then add systems to the respective stages to upload all necessary data to the gpu and draw it. Do know that other plugins only have integration for `bevy_render`, so you will need to add this integration to your own render plugin.

# Should you?

Probably not. `bevy_render` supports pretty much arbitrary render graphs. This includes rendering to a texture and then passing this texture to the next pass. I think it is possible to do what you want within the framework of `bevy_render`.

**Reply by Josh015** [reply]
I probably should have been more specific. I meant using a custom render graph, not replacing `bevy_render` entirely. With that in mind, are there any Bevy examples of rendering to a texture(s) and then rendering that to the screen?

**Reply by bjorn3** [reply]
> With that in mind, are there any Bevy examples of rendering to a texture(s) and then rendering that to the screen?

I don't think so. `bevy_render` could use some documentation. As far as I know `bevy_render` is also one of the most lacking areas of Bevy currently.

**Reply by FrankenApps** [reply]
I have seen [this](https://mrk.sed.pl/bevy-showcase/#render_to_texture). The code is [here](https://github.com/mrk-its/bevy/blob/render_to_texture/examples/3d/render_to_texture.rs)

**Comment by John-Nagle** [comment]
Can you pull out the Bevy renderer and run it without the rest of Bevy? Has anyone done that? Is there a sample 3D program? I've been using Rend3, and I need a plan B in case that project is abandoned.

**Reply by JMS55** [reply]
Not easily https://github.com/bevyengine/bevy/blob/main/crates/bevy_pbr/Cargo.toml#L16-L26. You could do something where you have a different thread copy values into Bevy's ECS every frame, but yeah not a great idea really.

**Reply by John-Nagle** [reply]
Yes, Bevy is built as an integrated system, not a game engine that uses a render library. 

Thanks anyway.
